File: src/state_scan.c
/*
 * state_scan.c
 *
 * Copyright (c) 2023 Thomas Buck (thomas@xythobuz.de)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * See <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <string.h>

#include "pico/stdlib.h"
#include "picowota/reboot.h"

#include "config.h"
#include "ble.h"
#include "models.h"
#include "mem.h"
#include "menu.h"
#include "state.h"
#include "state_workflow.h"
#include "state_volcano_run.h"
#include "state_volcano_conf.h"
#include "state_crafty.h"
#include "state_venty.h"
#include "state_scan.h"

static struct ble_scan_result results[BLE_MAX_SCAN_RESULTS] = {0};
static int result_count = 0;
static uint32_t auto_connect_time = 0;
static int auto_connect_idx = 0;

static void enter_cb(int selection) {
    int devs = 0;
    for (int i = 0; i < result_count; i++) {
        enum known_devices dev = models_filter_name(results[i].name);
        if (dev == DEV_UNKNOWN) {
            continue;
        }

        if (devs++ == selection) {
            if (dev == DEV_VOLCANO) {
                state_volcano_run_target(results[i].addr, results[i].type);
                state_wf_edit(false);
                state_switch(STATE_WORKFLOW);
            } else if (dev == DEV_CRAFTY) {
                state_crafty_target(results[i].addr, results[i].type);
                state_switch(STATE_CRAFTY);
            } else if (dev == DEV_VENTY) {
                state_venty_target(results[i].addr, results[i].type);
                state_switch(STATE_VENTY);
            }
            return;
        }
    }

    if (selection == devs) {
        state_switch(STATE_SETTINGS);
    } else if (selection == (devs + 1)) {
        state_switch(STATE_ABOUT);
    }
}

static void edit_cb(int selection) {
    int devs = 0;
    for (int i = 0; i < result_count; i++) {
        enum known_devices dev = models_filter_name(results[i].name);
        if (dev == DEV_UNKNOWN) {
            continue;
        }

        if (devs++ == selection) {
            if (dev == DEV_VOLCANO) {
                state_volcano_conf_target(results[i].addr, results[i].type);
                state_switch(STATE_VOLCANO_CONF);
            } else if (dev == DEV_CRAFTY) {
                state_crafty_target(results[i].addr, results[i].type);
                state_switch(STATE_CRAFTY);
            } else if (dev == DEV_VENTY) {
                state_venty_target(results[i].addr, results[i].type);
                state_switch(STATE_VENTY);
            }
            return;
        }
    }

    if (selection == devs) {
        state_switch(STATE_SETTINGS);
    } else if (selection == (devs + 1)) {
        state_switch(STATE_ABOUT);
    }
}

static void ota_cb(int selection) {
    UNUSED(selection);
    picowota_reboot(true);
}

void state_scan_enter(void) {
    menu_init(enter_cb, edit_cb, ota_cb, NULL);
    ble_scan(BLE_SCAN_ON);
}

void state_scan_exit(void) {
    ble_scan(BLE_SCAN_OFF);
    menu_deinit();
}

static void draw(struct menu_state *menu) {
    result_count = ble_get_scan_results(results, BLE_MAX_SCAN_RESULTS);

    int pos = 0, devs = 0;
    for (int i = 0; i < result_count; i++) {
        enum known_devices dev = models_filter_name(results[i].name);
        if (dev == DEV_UNKNOWN) {
            continue;
        }

        devs++;

        if (((devs - 1) < menu->off)
            || ((devs - 1 - menu->off) >= MENU_MAX_LINES)) {
            continue;
        }

#if defined(MENU_PREFER_VOLCANO) || defined(MENU_PREFER_CRAFTY)
        if (menu->selection < 0) {
#ifdef MENU_PREFER_VOLCANO
            if (dev == DEV_VOLCANO) {
                menu->selection = devs - 1;

                if (to_ms_since_boot(get_absolute_time()) <= VOLCANO_AUTO_CONNECT_WITHIN_MS) {
                    if (mem_data()->wf_auto_connect) {
                        auto_connect_time = to_ms_since_boot(get_absolute_time());
                        auto_connect_idx = i;
                    }
                }
            }
#endif // MENU_PREFER_VOLCANO
#ifdef MENU_PREFER_CRAFTY
            if (dev == DEV_CRAFTY) {
                menu->selection = devs - 1;
            }
#endif // MENU_PREFER_CRAFTY
        }
#endif // defined(MENU_PREFER_VOLCANO) || defined(MENU_PREFER_CRAFTY)

        if ((devs - 1) == menu->selection) {
            if ((auto_connect_time != 0) && (!menu_got_input)) {
                uint32_t now = to_ms_since_boot(get_absolute_time());
                uint32_t diff = now - auto_connect_time;
                pos += snprintf(menu->buff + pos, MENU_MAX_LEN - pos,
                                "%ld ",
                                (VOLCANO_AUTO_CONNECT_TIMEOUT_MS / 1000) - (diff / 1000));
            } else {
                pos += snprintf(menu->buff + pos, MENU_MAX_LEN - pos, "> ");
            }
        } else {
            pos += snprintf(menu->buff + pos, MENU_MAX_LEN - pos, "  ");
        }

        if (dev == DEV_VOLCANO) {
            pos += snprintf(menu->buff + pos, MENU_MAX_LEN - pos, "Volcano ");
        } else if (dev == DEV_CRAFTY) {
            pos += snprintf(menu->buff + pos, MENU_MAX_LEN - pos, "Crafty+ ");
        } else if (dev == DEV_VENTY) {
            pos += snprintf(menu->buff + pos, MENU_MAX_LEN - pos, "Venty ");
        }

        char info[32] = "";
        if (models_get_serial(dev, results[i].name,
                              results[i].data, results[i].data_len,
                              info, sizeof(info)) < 0) {
            strcpy(info, "-error-");
        }
        pos += snprintf(menu->buff + pos, MENU_MAX_LEN - pos, "%s\n", info);
    }

    menu->length = devs;

#if !defined(MENU_PREFER_VOLCANO) && !defined(MENU_PREFER_CRAFTY)
    if ((menu->selection < 0) && (menu->length > 0)) {
        menu->selection = 0;
    }
#endif // !defined(MENU_PREFER_VOLCANO) && !defined(MENU_PREFER_CRAFTY)

    ADD_STATIC_ELEMENT("Settings");
    ADD_STATIC_ELEMENT("About");
}

void state_scan_run(void) {
    menu_run(draw, false);

    if ((auto_connect_time != 0) && (!menu_got_input)) {
        uint32_t now = to_ms_since_boot(get_absolute_time());
        if ((now - auto_connect_time) >= VOLCANO_AUTO_CONNECT_TIMEOUT_MS) {
            state_volcano_run_target(results[auto_connect_idx].addr,
                                     results[auto_connect_idx].type);
            state_wf_edit(false);
            state_switch(STATE_WORKFLOW);

            auto_connect_time = 0;
        }
    }
}

File: test_micropython/state_scan.py
#!/usr/bin/env python3

# ----------------------------------------------------------------------------
# Copyright (c) 2023 Thomas Buck (thomas@xythobuz.de)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# See <http://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------

import uasyncio as asyncio
from scan import ble_scan
import time

class StateScan:
    def __init__(self, lcd):
        self.lcd = lcd

        self.lock = asyncio.Lock()

    def enter(self, val = None):
        n = None
        self.results = []
        self.current = None
        self.menuOff = 0
        self.scanner = asyncio.create_task(self.scan())

    def exit(self):
        self.scanner.cancel()

        if self.lock.locked():
            self.lock.release()

        self.lcd.store_brightness()

        return self.results[self.current][4]

    async def scan(self):
        while True:
            new = await ble_scan(None, None, 0.25)

            for n in new:
                name = n.name()
                mac = n.device.addr_hex()
                rssi = n.rssi
                value = [name, mac, rssi, time.time(), n]

                async with self.lock:
                    found = False
                    for i in range(0, len(self.results)):
                        if self.results[i][1] == mac:
                            found = True
                            self.results[i][0] = name
                            self.results[i][2] = rssi
                            self.results[i][3] = time.time()
                            self.results[i][4] = n
                            break

                    if found == False:
                        self.results.append(value)

    def draw_list(self):
        if len(self.results) <= 0:
            self.lcd.textC("No devices found yet", int(self.lcd.width / 2), int(self.lcd.height / 2), self.lcd.white)
            return

        for i, d in enumerate(self.results):
            if i < self.menuOff:
                continue

            off = (i - self.menuOff) * 25 + 30
            if off >= (self.lcd.height - 10):
                break

            selection = "  "
            if self.current == i:
                selection = "->"
            name, mac, rssi, timeout, device = self.results[i]
            age = int(time.time() - timeout)
            s1 = "{}: {}".format(i + 1, name)
            s2 = "{} [{}] {} {}".format(selection, mac, rssi, age)

            c1 = self.lcd.white
            if name == "S&B VOLCANO H":
                c1 = self.lcd.green
                if self.current == None:
                    self.current = i
            elif name == "STORZ&BICKEL":
                c1 = self.lcd.yellow
            elif self.current == i:
                c1 = self.lcd.red
            c2 = self.lcd.white
            if self.current == i:
                c2 = self.lcd.red

            self.lcd.hline(0, off - 3, self.lcd.width, self.lcd.blue)
            self.lcd.text(s1, 0, off + 2, c1)
            self.lcd.text(s2, 0, off + 12, c2)

    async def draw(self):
        self.lcd.text("Scanning for Bluetooth devices", 0, 10, self.lcd.red)

        keys = self.lcd.buttons()

        async with self.lock:
            if keys.once("enter") or keys.once("a"):
                if self.current < len(self.results):
                    return 2
            elif keys.once("up"):
                if self.current == None:
                    self.current = len(self.results) - 1
                else:
                    self.current -= 1
            elif keys.once("down"):
                if self.current == None:
                    self.current = 0
                else:
                    self.current += 1
            elif keys.once("x"):
                return 10
            elif keys.held("left"):
                v = self.lcd.curr_brightness - 0.05
                if v < 0.05:
                    v = 0.05
                self.lcd.brightness(v)
            elif keys.held("right"):
                self.lcd.brightness(self.lcd.curr_brightness + 0.05)
            elif keys.once("y"):
                return 0

            if self.current != None:
                while self.current < 0:
                    self.current += len(self.results)
                while self.current >= len(self.results):
                    self.current -= len(self.results)
                while self.current < self.menuOff:
                    self.menuOff -= 1
                while self.current >= (self.menuOff + int((self.lcd.height - 30 - 10) / 25)):
                    self.menuOff += 1

            # remove entries after timeout
            self.results = [x for x in self.results if (time.time() - x[3]) < 10.0]

            # filter out incompatible devices
            self.results = [x for x in self.results if (x[0] != None) and (("S&B" in x[0]) or ("STORZ&BICKEL" == x[0]))]

            if self.current != None:
                if self.current >= len(self.results):
                    self.current = len(self.results) - 1

            self.draw_list()

        w = int(self.lcd.width * self.lcd.curr_brightness)
        self.lcd.hline(0, 24, w, self.lcd.white)

        return -1
